### ISSUE TITLE
Properly inject a runtime

### ISSUE THREAD

**Issue by IngwiePhoenix** [webpack/webpack#1001]
I am working on a loader for [OJ](https://github.com/musictheory/oj) that compiles `.oj` files to JS. So far, this works very nicely especially since the compiler doesn't mind an async loader.

There is one catch that I came across that is making things a bit complicated_

When I inject the runtime via the following header:

```javascript
    var header = [
        "var oj = require('!!ojc/src/runtime');",
        "global.oj = global['$oj_oj'] = oj;"
    ].join("\n");
```

I get back a snippet in my output file that looks like:

```javascript
/***/ function(module, exports, __webpack_require__) {
	/* WEBPACK VAR INJECTION */(function(global) {
            // Within here we have the runtime. Runtime calls itself like:
            ;(function(){ /* snip */ }).call(this);
	/* WEBPACK VAR INJECTION */}.call(exports, (function() { return this; }())))
/***/ }
```

The problem with this injection is, that `exports` is transformed  into the `this` of the function - and further, that internal function calls its first argument `global`, effectively shaddowing the real global (`window` in a browser). And I can't quite recognize what `(function() { return this; }())` will return... and I doubt it's the real global, which is **mandatory**.

My questions are:

- How do I properly inject this runtime?
- Is there a way to only inject my header once into the global space? As in, reducing the duplication of code. `$oj_oj` is expected to keep its state, so I only want to assign it once, and only once.

Kind regards, Ingwie!

**Comment by IngwiePhoenix** [comment]
Turns out that the seemingly shaddowed `global` is indeed the real global. I just made a quick and dirty test:

```javascript
webpackJsonp([0,1],[
/* 0 */
/***/ function(module, exports, __webpack_require__) {

	console.log("meep.");
	__webpack_require__(1);
	//require("./lib.oj");


/***/ },
/* 1 */
/***/ function(module, exports, __webpack_require__) {

	/* WEBPACK VAR INJECTION */(function(global) {console.log("This:");
	console.log(this);
	console.log("Global:");
	console.log(global);

	/* WEBPACK VAR INJECTION */}.call(exports, (function() { return this; }())))

/***/ }
]);                                                                                                                                                                                                        
```

The question now is, how do I run my code injection only once?

**Comment by sokra** [comment]
Don't use `global`. Don't add anything to the global scope.

To add a runtime best do it like:

``` js
var loaderUtils = require("loader-utils");
module.exports = function() {
  return "var runtime = require(" + loaderUtils.stringifyRequest(this, "!" + require.resolve("ojc/src/runtime")) + ");";
}
```

Add ojc to `peerDependencies`.

Also check these loader guidelines: https://github.com/webpack/docs/wiki/how-to-write-a-loader#guidelines

**Comment by IngwiePhoenix** [comment]
OJ requires to add `oj` and `$oj_oj` to the global object, since the compiled code reffers to it. For instance:

```javascript
var Bar = $oj_oj._registerClass({ $oj_c_Bar:1 }, { $oj_c_Foo:1 }, function($oj_s, $oj_m) { function Bar() { $oj_oj._cls.$oj_c_Foo.call(this);this.constructor = Bar;this.$oj_id = ++$oj_oj._id;}
$oj_s.$oj_f_isBard = function() { return true; }
return Bar;});
```

The original for this code is:

```objective-c
@implementation Bar:Foo {}
+(BOOL)isBard { return true; }
@end
```

So in order to allow one class to extend another, the global `$oj_oj` object needs to retain its state - the added classes and such. the `oj` object is used to run the code, i.e. `oj.sendMsg(...)`.

Therefore the runtime must be injected in global scope. In fact, it tries to inject itself:

https://github.com/musictheory/oj/blob/master/src/runtime.js#L366

I thought about looking into making this specific part a plugin; so that it injects the runtime, whilst the loader does the compiling. What do you think?

**Comment by IngwiePhoenix** [comment]
I just looked through the (huge!) plugin API. But I can not determine where I would hook into in order to inject the runtime. Since I am using the `commonsJunk` plugin, it makes two separate junks - so I only want to inject the runtime exactly once, preferably into the commons junk.

- Which part do I hook into?
- Is my plugin persistent enough to retain state - i.e. if I set `this.isInjected`, will this stay that way?

**Comment by SimenB** [comment]
@IngwiePhoenix What about https://github.com/webpack/imports-loader?

**Comment by IngwiePhoenix** [comment]
@SimenB I am not sure with this loader - will it retain state, load my script only once and into global scope... I will be using this loader for another section of my code, but I have to let the runtime run in the global scope and be only added once.

That is why I was looking into maybe using a plugin, since i have to ensure that the runtime exists only once... Do you have any idea how I could do that? My loader, that takes OJ code and compiles it, works. I can even make sourcemaps now. But I am unable to get the runtime injected properly :/

**Comment by sokra** [comment]
> That is why I was looking into maybe using a plugin, since i have to ensure that the runtime exists only once...

You don't need a plugin. Loader is enough.

`require("xxx") === require("xxx")`.

`var $oj_oj = require(...);` should solve all problems.

**Comment by IngwiePhoenix** [comment]
I just tried that, and the result:

    Uncaught ReferenceError: Foo is not defined

The output code is here:  https://gist.github.com/IngwiePhoenix/e3743bcd3bc6bbdcf43f

It appears that the `var`s in the generated code need to go into a reachable area - but I am not sure how.

**Comment by IngwiePhoenix** [comment]
Pinned it... im going to go and eat a broomstick INCLUDING the broom.

Turns out that OJ optimizes previously used variable names. Here is what it ACTUALLY writes, when the class is in the same file:

    $oj_oj.msgSend($oj_oj._cls.$oj_c_Foo,{ $oj_f_bard: 1 })

Since the `Foo` interface did not exist in the other file, the `Foo` constant was not resolved, hence:

    $oj_oj.msgSend(Foo,{ $oj_f_bard: 1 })

So yeah. This is an issue on OJ, and one that I have to figure out, somehow...probably has to do with it's optimizer.

Thanks for all of your help, I finally figured it out! Oh man. ._. Maybe @iccir has a solution for this, as he is the maintainer of OJ.

Once this works with WebPack, ill make sure to let you know so it can be added to the list of loaders! :)

Kind regards, Ingwie
